File: src/main/java/org/micoli/php/symfony/messenger/usage/MessengerFindUsagesHandlerFactory.java
package org.micoli.php.symfony.messenger.usage;

import com.intellij.find.findUsages.FindUsagesHandler;
import com.intellij.find.findUsages.FindUsagesHandlerFactory;
import com.intellij.openapi.application.ReadAction;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import com.jetbrains.php.lang.psi.elements.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.micoli.php.symfony.messenger.service.MessengerService;

public class MessengerFindUsagesHandlerFactory extends FindUsagesHandlerFactory {

    @Override
    public boolean canFindUsages(@NotNull PsiElement element) {
        return ReadAction.compute(() -> {
            MessengerService messengerService = MessengerService.getInstance(element.getProject());
            boolean isATrigger = false;
            if (element instanceof Method method) {
                if (messengerService.isHandlerMethod(method.getName())) {
                    isATrigger = true;
                }
            }
            if (element instanceof PhpClass) {
                isATrigger = true;
            }
            if (!isATrigger) {
                return false;
            }

            PhpClass phpclass = getMessageClass(messengerService, element);
            return (phpclass != null);
        });
    }

    @Override
    public @Nullable FindUsagesHandler createFindUsagesHandler(
            @NotNull PsiElement element, boolean forHighlightUsages) {
        return ReadAction.compute(() -> {
            if (canFindUsages(element)) {
                return new MessengerFindUsagesHandler(element);
            }
            return null;
        });
    }

    public static PhpClass getMessageClass(MessengerService messengerService, PsiElement element) {
        return ReadAction.compute(() -> {
            if (element instanceof PhpClass phpClass) {

                if (messengerService.isMessageClass(phpClass)) {
                    return phpClass;
                }
                if (messengerService.isHandlerClass(phpClass)) {
                    return messengerService.getHandledMessage(phpClass);
                }
            }

            PhpClass handlerClass = PsiTreeUtil.getParentOfType(element, PhpClass.class);
            if (handlerClass == null) {
                return null;
            }
            return messengerService.getHandledMessage(handlerClass);
        });
    }
}

File: src/main/kotlin/org/micoli/php/symfony/messenger/usage/MessengerFindUsagesHandlerFactory.kt
package org.micoli.php.symfony.messenger.usage

import com.intellij.find.findUsages.FindUsagesHandler
import com.intellij.find.findUsages.FindUsagesHandlerFactory
import com.intellij.openapi.application.ReadAction
import com.intellij.psi.PsiElement
import com.intellij.psi.util.PsiTreeUtil
import com.jetbrains.php.lang.psi.elements.Method
import com.jetbrains.php.lang.psi.elements.PhpClass
import org.micoli.php.symfony.messenger.service.MessengerService

class MessengerFindUsagesHandlerFactory : FindUsagesHandlerFactory() {
    override fun canFindUsages(element: PsiElement): Boolean {
        return ReadAction.compute<Boolean, RuntimeException?> {
            val messengerService: MessengerService = MessengerService.getInstance(element.project)
            var isATrigger = false
            if (element is Method) {
                if (messengerService.isHandlerMethod(element.name)) {
                    isATrigger = true
                }
            }
            if (element is PhpClass) {
                isATrigger = true
            }
            if (!isATrigger) {
                return@compute false
            }

            val phpclass: PhpClass? = getMessageClass(messengerService, element)
            (phpclass != null)
        }
    }

    override fun createFindUsagesHandler(
        element: PsiElement,
        forHighlightUsages: Boolean
    ): FindUsagesHandler? {
        return ReadAction.compute<MessengerFindUsagesHandler, RuntimeException?> {
            if (canFindUsages(element)) {
                return@compute MessengerFindUsagesHandler(element)
            }
            null
        }
    }

    fun getMessageClass(messengerService: MessengerService, element: PsiElement?): PhpClass? {
        return ReadAction.compute<PhpClass?, RuntimeException?> {
            if (element is PhpClass) {
                if (messengerService.isMessageClass(element)) {
                    return@compute element
                }
                if (messengerService.isHandlerClass(element)) {
                    return@compute messengerService.getHandledMessage(element)
                }
            }
            val handlerClass =
                PsiTreeUtil.getParentOfType(element, PhpClass::class.java) ?: return@compute null
            return@compute messengerService.getHandledMessage(handlerClass)
        }
    }
}
